lib/codecs/src/decoding/format/protobuf.rs

```diff
@@ -0,0 +1,311 @@
+                        Kind::json().or_timestamp(), // TODO: create kind::protobuf?
+                        Some("timestamp"),
+                    );
+                }
+                definition
+            }
+            LogNamespace::Vector => {
+                schema::Definition::new_with_default_metadata(Kind::json(), [log_namespace])
+                // TODO: create kind::protobuf?
+            }
+        }
+    }
+}
+
+impl ProtobufDeserializerConfig {
+    /// Creates a new `ProtobufDeserializerConfig`.
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
+
+/// Deserializer that builds `Event`s from a byte frame containing protobuf.
+#[derive(Debug, Clone)]
+pub struct ProtobufDeserializer {
+    message_descriptor: MessageDescriptor,
+}
+
+impl ProtobufDeserializer {
+    /// Creates a new `ProtobufDeserializer`.
+    pub fn new(desc_file: String, message_type: String) -> Self {
+        // TODO: handle 'expect' in a better way
+        let b = fs::read(desc_file.clone())
+            .expect(&format!("Failed to open protobuf desc file '{desc_file}'"));
+        let pool = DescriptorPool::decode(b.as_slice())
+            .expect(&format!("Failed to parse protobuf desc file '{desc_file}'"));
+        Self {
+            message_descriptor: pool.get_message_by_name(&message_type).expect(&format!(
+                "The message type '{message_type}' could not be found in '{desc_file}'"
+            )),
+        }
+    }
+}
+
+impl Deserializer for ProtobufDeserializer {
+    fn parse(
+        &self,
+        bytes: Bytes,
+        log_namespace: LogNamespace,
+    ) -> vector_common::Result<SmallVec<[Event; 1]>> {
+        let dynamic_message = DynamicMessage::decode(self.message_descriptor.clone(), bytes)
+            .map_err(|error| format!("Error parsing protobuf: {:?}", error))?;
+
+        let proto_vrl = to_vrl(
+            prost_reflect::Value::Message(dynamic_message),
+            &prost_reflect::Kind::Message(self.message_descriptor.clone()),
+        )?;
+        let mut event = Event::Log(LogEvent::from(proto_vrl));
+        let event = match log_namespace {
+            LogNamespace::Vector => event,
+            LogNamespace::Legacy => {
+                let timestamp = Utc::now();
+
+                if let Some(timestamp_key) = log_schema().timestamp_key() {
+                    let log = event.as_mut_log();
+                    if !log.contains((PathPrefix::Event, timestamp_key)) {
+                        log.insert((PathPrefix::Event, timestamp_key), timestamp);
+                    }
+                }
+
+                event
+            }
+        };
+
+        Ok(smallvec![event])
+    }
+}
+
+impl From<&ProtobufDeserializerConfig> for ProtobufDeserializer {
+    fn from(config: &ProtobufDeserializerConfig) -> Self {
+        Self::new(config.desc_file.clone(), config.message_type.clone())
+    }
+}
+
+fn to_vrl(
+    prost_reflect_value: prost_reflect::Value,
```

Comment:
I played around with this locally, it is possible to pass `prost_reflect_value` by reference. Maybe the bytes branch will need a clone but feel free to explore it yourself.

Reply:
I wanted to avoid unneeded copy and allocations, I tried to "steal" all the resources from the value itself.
you can see the difference in to_vrl_by_ref in my cargo bench demo.
please take a look at the demo info below and let me know what you prefer.